### ISSUE TITLE
Create `rc_replica_truncate` POSIX-like `truncate(2)` API

### ISSUE THREAD

**Issue by korydraughn** [irods/irods#7104]
- [x] main
- [x] 4-3-stable
---
## Investigation / Refactor
Clients such as NFSRODS need to be able to truncate (resize) data objects. There are UNIX tools which rely on this behavior. To support this, clients must simulate the operation. This results in degraded performance.

The current implementation of `rxDataObjTruncate` is partially there. All that's needed is for it to honor logical locking.

Given that truncation is a destructive operation, the data object being targeted should only be modified if it has a replica status of _good_ or _stale._ The truncation operation must fail immediately if the data object has a replica status of _intermediate._

This issue is motivated by the following conversation.
- https://github.com/irods/irods_client_nfsrods/issues/178

**Comment by alanking** [comment]
A set of requirements has formed... I will number them for easy reference in any further discussion.

1. Introduce 1 or 2 new API plugins to mimic `truncate` and `ftruncate`. These may or may not deprecate `rxDataObjTruncate`. We will try to follow `truncate` behavior as closely as possible.
2. Only one replica can be truncated at a time. This implies support for targeting a replica number and a root resource. If none is supplied, the default resource will be targeted. Hierarchy resolution will determine which replica will be targeted. I propose that it operates on a "write" operation. This differs from the existing `rxDataObjTruncate` which truncates all replicas to the specified size.
3. Open objects can be targeted for truncate and only the intermediate replica can be truncated.
4. `truncate` cannot create files, so these APIs will not create data objects or replicas.
5. We need to write extensive unit tests to make sure things are working as intended because this operation is destructive to data.
6. There may or may not be a new iCommand to go with the new API (`itrunc` or `itruncate`). This is definitely not a priority item.

**Comment by alanking** [comment]
This may require implementing a new resource plugin operation for `ftruncate`. The unixfilesystem resource implementation of the truncate resource plugin operation literally just calls `truncate`. In order to support `ftruncate` we would have to...

a. modify the signature of the operation to accept a file descriptor
b. smuggle it in via a keyword
c. create a different operation for `ftruncate`

I think implementing a separate plugin operation (c) is the cleanest choice.

**Comment by korydraughn** [comment]
I agree.

**Comment by korydraughn** [comment]
Noting here this will affect any plugins which track changes in data objects (e.g. logical quotas).

**Comment by alanking** [comment]
Do we expect the new truncate APIs to behave like `truncate` and `ftruncate` in the way that they handle output and error codes? 

Historically, when an iRODS API returns an error, it typically returns the error code as the return value (an integer) and maybe puts a message on the rError stack. When `truncate` has an error, it returns -1 and sets `errno` to some value.

My plan with these APIs was to return -1 like `truncate` and `ftruncate` and include the iRODS error code and a message in a JSON structure like this:
```json
{
    // An integer converted from an iRODS error code, which is a rodsLong_t. 0 on success.
    "irods_error_code": -808000,
    // Some message made up by the API if it has something useful to say. Usually empty on success.
    "message": "No rows found!"
}
```
Any thoughts or alternative ideas encouraged :)

**Comment by trel** [comment]
what is the argument for returning -1 on error?    are we trying to meet some compatibility or expectation?

why is this new API different from our others?

**Comment by korydraughn** [comment]
I don't think returning -1 helps us here very much. I think the benefit has to be pretty huge for us to deviate from how we normally report errors. With the proposed designed, applications have to parse a JSON string to know what the issue is.

Do we have other POSIX-like APIs which do what is being proposed?
What are the error cases for `rsDataObjruncate`?

**Comment by alanking** [comment]
Mostly, I was trying to conform to the behavior of `truncate` and `ftruncate`. I realize this isn't entirely possible due to the fact that there's no global `errno` concept in our APIs (at least that I know of). Putting the error information in the output string seemed to strike a certain balance as far as trying to match the behavior the best I could... But maybe that's not the goal? 

In any case, this is indeed a deviation from the existing POSIX-like iRODS APIs, but we have gone on record as saying we don't like those very much. ;) Just experimenting with whether it would be useful to move to be "closer" in usage to those APIs. I suppose that would entail trying to match the signatures, too, which we obviously are not able to do.

Sounds like maybe we're not so into the idea of trying to match the POSIX APIs. I think I agree that returning the error code directly is a lot easier to program against.

**Comment by trel** [comment]
yeah, i think the ship has sailed on our API call and response format.   we'd have to change a lot to change it.

HTTP API as a wrapper is kind of solving some of that pain/accessibility.

**Comment by alanking** [comment]
Okay, well, with that, now I need to confirm some other things before I write anything else.

Here was my plan for the function signature, cuing off the API plugin template:
```c
int rc_replica_truncate(struct RcComm* _comm, const char* _input, char** _output);
```

And here is the input structure:
```json
{
    "path": "<string>", 
    "length": <integer>,
    // Optional. Incompatible with replica_number.
    "target_resource": "<string>", 
    // Optional. Incompatible with target_resource.
    "replica_number": <integer>
}
```

If we wanted to make it more like the POSIX API (https://pubs.opengroup.org/onlinepubs/009604599/functions/truncate.html)...
```c
int truncate(const char *path, off_t length); // By all accounts, off_t is mysterious and vaguely defined and is best to be avoided.
```
...the signature would look something more like this:
```c
int rc_replica_truncate(struct RcComm* _comm, const char* _path, int _length, KeyValPair* _options);
```

Where `_options` would contain the `target_resource` and `replica_number` options (and any other options that might come along). This could even come with its own input/output struct(s), but I am pretty strongly against continuing that tradition. This is less friendly to non-C/C++ clients, but more in line with how the other POSIX-like iRODS APIs look.

**Comment by korydraughn** [comment]
I like the name, `rc_replica_truncate`. It's clear it's about a single replica. Although, as I type this, is the use of `rcDataObj` incorrect? _Just wondering._

If we go with JSON, the function becomes future proof at the expense of more parsing.
If we go with the traditional PackStruct format, parsing is handled for us.
Do we suspect potential changes/additions to the truncate API?
_This is really a performance question. That is, how much overhead does parsing JSON introduce._

Is the trailing comment about `off_t` being vaguely defined coming from you or POSIX?

> Where `_options` would contain the `target_resource` and `replica_number` options (and any other options that might come along). This could even come with its own input/output struct(s), but I am pretty strongly against continuing that tradition. This is less friendly to non-C/C++ clients, but more in line with how the other POSIX-like iRODS APIs look.

I don't think you can actually avoid the input/output setup due to how information is passed to the iRODS server. The rc function signature is fine as a convenience interface, but underneath, it's required to use some structure for the input and output.

At the moment, we haven't defined how to ship the implementation for convenience functions so C/C++ clients are stuck with `procApiRequest`.

**Comment by korydraughn** [comment]
I agree. Let's start with a `DataObjInp` for accepting input. As you said, that type covers quite a number of things and is familiar to users.

We can discuss the name of the API as the real work approaches completion.

Does that cover the minimum requirements?

**Comment by alanking** [comment]
I think so... I'll write back here if I think of anything else.

**Comment by alanking** [comment]
Okay, so, through some experimentation with an initial implementation, it seems we may not be able to avoid changes to the server in order for this to work the way we want it to. It comes down to truncate needing its own voting logic rather than piggy-backing on read or write.

For starters, truncate is most similar to a write (without the open, of course) so it follows that the hierarchy resolution would be for `irods::WRITE_OPERATION`. However, for reasons which I won't get into here, `irods::WRITE_OPERATION` votes higher for stale replicas. A truncate usually precedes other modifying operations, very commonly an *append*. It would make more sense to update a good replica rather than a stale one, so it follows that truncate should prefer good replicas.

While `irods::OPEN_OPERATION` votes higher for good replicas, it is also supposed to represent a read, which is not what truncate does. The proposal is to add a new voting operation for truncate (same as is being proposed for rename in #4922).

Initially, we would not expose this as a separately configurable weight on the passthru resource. The "write" weight should affect truncates. The truncate voting operation would just vote higher for good replicas.

**Comment by alanking** [comment]
We have determined that the way voting for write works is a flawed design, and we will need to change it per #7476. As such, truncate can vote on write operations and will work as expected.

**Comment by alanking** [comment]
Got a progress update here...

We addressed the community at the March 2024 Technology Working Group and received no feedback regarding usage of the existing API. As such, I have created #7555 to deprecate the existing API. This issue will relate to the creation of a *new* API which will achieve the desired behavior.

**Comment by trel** [comment]
consider renaming this issue with the name of the new endpoint

**Comment by alanking** [comment]
I have another question regarding the output of this API.

The current signature that I have implemented for this function looks almost identical to `rcDataObjTruncate`:
```
int rc_replica_truncate(RcComm* _comm, DataObjInp* _inp);
```
It will behave like our APIs always have which is to say that the `_inp` parameter will be advertised as an "in" when in reality is an in/out parameter. Should we continue in this grand tradition, or should we make `_inp` a `const DataObjInp*` and provide an output parameter to communicate things such as truncated replica information, error messages, etc.?

**Comment by trel** [comment]
seems reasonable/good.    we have no other precedent for an out-exclusive parameter?

**Comment by alanking** [comment]
There is precedent for both in-only and out-only parameters in our APIs. Historically, however, many have taken advantage of the in parameter to also communicate information back to the client and has even become expected behavior for some clients. Most notably, PRC and I believe one or more REPs depend on the hierarchy of the source and destination replica to be returned through the input structure: #6100.

**Comment by trel** [comment]
Okay, yeah, same vote.  Seems good to separate them.   Any downsides?

**Comment by alanking** [comment]
If we use the in/out structure, no additional parsing or memory management is required on the part of the client. It's not the worst, but I prefer separated inputs and outputs. This is the current implementation.

If we use a JSON string for the output parameter, we can return whatever information we think we might need. The downside is having to parse the output on the client side and manage the memory of the returned string. This is my favorite option.

If we use an existing iRODS structure (such as a `KeyValPair**`), this is a little clunky, but the clients and serializers will already be familiar with it. The downside is that the interface is not as flexible. I am okay with this option if it can be made to work.

If we do a new kind of structure for the output, that's one more thing that needs to be serialized for REPs, one more thing for PackStruct to learn, and one more thing that clients will have to deal with. This is my least favorite option, even beneath using the input structure to communicate information back to the client.

**Comment by trel** [comment]
JSON string as out parameter seems... the cleanest and most flexible/future-proof.

**Comment by korydraughn** [comment]
A third parameter for outputting a json string sounds good/best. Marking the input struct const sounds good too.

**Comment by alanking** [comment]
Great, I'll put in an output parameter, then

**Comment by alanking** [comment]
Leaving a note for my future self and anybody else who might be interested... Seems that `truncate()` of open files by other processes is allowed in Linux, but can lead to some weird results due to in-memory file descriptor information:
https://unix.stackexchange.com/questions/649094/truncate-file-that-is-opened-in-another-procces

I'm thinking that we should not allow other processes to truncate an open replica even with a replica access token. Once a replica has been opened, we can provide an `ftruncate()` for use by any agent process with a file descriptor, which would require usage of a replica access token to open the replica.

**Comment by korydraughn** [comment]
> I'm thinking that we should not allow other processes to truncate an open replica even with a replica access token. Once a replica has been opened, we can provide an `ftruncate()` for use by any agent process with a file descriptor, which would require usage of a replica access token to open the replica.

Agreed. `ftruncate` is the answer.

**Comment by trel** [comment]
Double agree.  Please run away from that scenario.

**Comment by alanking** [comment]
I split off `ftruncate` in a new issue: https://github.com/irods/irods/issues/7559 It's going to be a bit of a lift as it turns out, so we should do that work separately, especially if this API satisfies the original use case (which it sounds like it may).

**Comment by alanking** [comment]
The initial implementation is now complete and will be included in the 4.3.2 server as a built-in API.

The `ftruncate`-like API will be handed in https://github.com/irods/irods/issues/7559

The microservice to accompany this API will be handled in #7577
